docs: Update terminationGracePeriod (#1544)
redpanda-data/documentation: docs/manage/kubernetes/rolling-upgrade.mdx

## Perform a rolling upgrade

A rolling upgrade involves putting a broker into [maintenance mode](../../node-management), upgrading the broker, taking the broker out of maintenance mode, and then repeating the process on the next broker in the cluster. Placing brokers into maintenance mode ensures a smooth upgrade of your cluster while reducing the risk of interruption or degradation in service.

> When a broker is placed into maintenance mode, it reassigns its partition leadership to other brokers for all topics that have a replication factor greater than one. Reassigning partition leadership involves _draining_ leadership from the broker and _transferring_ that leadership to another broker. If you have topics with `replication.factor=1`, and if you have sufficient disk space, Redpanda recommends temporarily increasing the replication factor. This can help limit outages for these topics during the rolling upgrade. Do this before the upgrade to make sure there's time for the data to replicate to other brokers. For more information, see [Change topic replication factor](../../data-migration/#change-topic-replication-factor).

Reviewer: ```suggestion
When a broker is placed into maintenance mode, it reassigns its partition leadership to other brokers for all topics that have a replication factor greater than one. Reassigning partition leadership involves _draining_ leadership from the broker and _transferring_ that leadership to another broker. If you have topics with `replication.factor=1` and sufficient disk space, Redpanda recommends temporarily increasing the replication factor. This can help limit outages for these topics during the rolling upgrade. Do this before the upgrade to ensure there's time for the data to replicate to other brokers. For details, see [Change topic replication factor](../../data-migration/#change-topic-replication-factor).
```

To perform a rolling upgrade:

1. [Deploy an upgraded StatefulSet](#deploy-an-upgraded-statefulset) with your desired Redpanda version.
2. [Upgrade and restart each broker](#upgrade-the-brokers) separately, one after the other.

:::info

> Redpanda Data does not recommend using the `kubectl rollout restart` command to perform rolling upgrades. Although the chart's `preStop` lifecycle hook puts the broker into maintenance mode before a Pod is deleted, the `terminationGracePeriod` may not be long enough to allow maintenance mode to finish. If maintenance mode does not finish before the Pod is deleted, you may lose data. After the `terminationGracePeriod`, the container is forcefully stopped using a `SIGKILL` command.

Reviewer: hmmm--next we talk, I'd like to understand how maintenance mode "finishes"--isn't this just taking a broker offline while you upgrade? How can/does that "finish"? Are you referring to how long it takes to come back online after you take it out of maintenance mode?

Author: When a broker is in maintenance mode, it's partitions are moved to other brokers in the cluster to avoid data loss.

> You can use the [`statefulset.terminationGracePeriodSeconds`](../../../reference/redpanda-helm-spec#statefulsetterminationgraceperiodseconds) setting to configure the `terminationGracePeriod`. However, it is difficult to determine the necessary value. In common cases 30 seconds should be long enough. For large clusters 90 seconds should be long enough. You can test different values in a development environment.

Reviewer 2: Is this for someone not using helm?

Author: This is for Helm. I asked @RafalKorepta and he said this is still the advice

Reviewer 3: If the user upgrades later on that value will be reset. I think this is what @joejulian is getting at.

Author: @alejandroEsc @joejulian isn't our advice to include all overrides when upgrading? So the value shouldn't be reset?

> You can use the [`statefulset.terminationGracePeriodSeconds`](../../../reference/redpanda-helm-spec#statefulsetterminationgraceperiodseconds) setting to configure the `terminationGracePeriod`. However, it is difficult to determine the necessary value. In common cases 30 seconds should be long enough. For large clusters 90 seconds should be long enough. You can test different values in a development environment.

Reviewer: ```suggestion
Use the [`statefulset.terminationGracePeriodSeconds`](../../../reference/redpanda-helm-spec#statefulsetterminationgraceperiodseconds) setting to configure the `terminationGracePeriod`. However, it is difficult to determine the necessary value. In common cases 30 seconds should be sufficient. For large clusters, 90 seconds should be sufficient. You can test different values in a development environment.
```

To perform a rolling upgrade:

1. [Deploy an upgraded StatefulSet](#deploy-an-upgraded-statefulset) with your desired Redpanda version.
2. [Upgrade and restart each broker](#upgrade-the-brokers) separately, one after the other.

:::info
Redpanda Data does not recommend using the `kubectl rollout restart` command to perform rolling upgrades. Although the chart's `preStop` lifecycle hook puts the broker into maintenance mode before a Pod is deleted, the `terminationGracePeriod` may not be long enough to allow maintenance mode to finish. If maintenance mode does not finish before the Pod is deleted, you may lose data. After the `terminationGracePeriod`, the container is forcefully stopped using a `SIGKILL` command.

> Use the [`statefulset.terminationGracePeriodSeconds`](../../../reference/redpanda-helm-spec#statefulsetterminationgraceperiodseconds) setting to configure the `terminationGracePeriod`.
> 
>  :::note
>  It can be a challenge to determine the necessary value for the `terminationGracePeriod`. In common cases, 30 seconds should be sufficient. For large clusters, 90 seconds should be sufficient. You can test different values in a development environment.
>  :::

Author: ```suggestion
It can be a challenge to determine the necessary value for the `terminationGracePeriod`. In common cases, 30 seconds should be sufficient. For large clusters, 90 seconds should be sufficient. You can test different values in a development environment. To configure the `terminationGracePeriod`, use the [`statefulset.terminationGracePeriodSeconds`](../../../reference/redpanda-helm-spec#statefulsetterminationgraceperiodseconds) setting.
```

## Rollbacks

> If something does not go as planned during a rolling upgrade, you can roll back to the original version as long as you have not upgraded every broker. The StatefulSet uses the `RollingUpdate` strategy by default in [`statefulset.updateStrategy.type`](../../../reference/redpanda-helm-spec#statefulsetupdatestrategytype), which means all Pods in the StatefulSet are restarted in reverse-ordinal order. For more details, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#update-strategies).

Reviewer: ```suggestion
If something does not go as planned during a rolling upgrade, you can roll back to the original version as long as you have not upgraded every broker. The StatefulSet uses the `RollingUpdate` strategy by default in [`statefulset.updateStrategy.type`](../../../reference/redpanda-helm-spec#statefulsetupdatestrategytype), which means all Pods in the StatefulSet are restarted in reverse-ordinal order. For more details, see [Kubernetes Update Strategies](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#update-strategies).
```

> If something does not go as planned during a rolling upgrade, you can roll back to the original version as long as you have not upgraded every broker. The StatefulSet uses the `RollingUpdate` strategy by default in [`statefulset.updateStrategy.type`](../../../reference/redpanda-helm-spec#statefulsetupdatestrategytype), which means all Pods in the StatefulSet are restarted in reverse-ordinal order. For more details, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#update-strategies).

Reviewer: ```suggestion
If something does not go as planned during a rolling upgrade, you can roll back to the original version as long as you have not upgraded every broker. The StatefulSet uses the `RollingUpdate` strategy by default in [`statefulset.updateStrategy.type`](../../../reference/redpanda-helm-spec#statefulsetupdatestrategytype), which means all Pods in the StatefulSet are restarted in reverse-ordinal order. For details, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#update-strategies).
```
